src/sentry/snuba/metrics_layer/query.py

```diff
@@ -77,18 +77,111 @@ def __init__(self) -> None:
         self.reverse_mappings: dict[int, str] = dict()
 
 
+def bulk_run_query(requests: list[Request]) -> list[Mapping[str, Any]]:
+    """
+    Entrypoint for executing a list of metrics queries in Snuba.
+
+    This function is used to execute multiple metrics queries in a single request.
+    """
+    queries = []
+    for request in requests:
+        request, start, end = _setup_metrics_query(request)
+        queries.append([request, start, end])
+
+    logging_tags = {"referrer": request.tenant_ids["referrer"] or "unknown", "lang": "mql"}
+
+    for q in queries:
+        q[0], reverse_mappings, mappings = _resolve_metrics_query(q[0], logging_tags)
+        q.extend([reverse_mappings, mappings])
+
+    try:
+        snuba_results = bulk_snuba_queries(
+            [q[0] for q in queries],
+            queries[0][0].tenant_ids["referrer"],
+            use_cache=True,
+        )
+    except Exception:
+        metrics.incr(
+            "metrics_layer.query",
+            tags={**logging_tags, "status": "query_error"},
+        )
+        raise
+
+    for idx, snuba_result in enumerate(snuba_results):
+        request, start, end, reverse_mappings, mappings = queries[idx]
+        metrics_query = request.query
+
+        snuba_result = convert_snuba_result(
+            snuba_result,
+            reverse_mappings,
+            request.dataset,
+            metrics_query.scope.use_case_id,
+            metrics_query.scope.org_ids[0],
+        )
+
+        # If we normalized the start/end, return those values in the response so the caller is aware
+        results = {
+            **snuba_result,
+            "modified_start": start,
+            "modified_end": end,
+            "indexer_mappings": mappings,
+        }
+
+        snuba_results[idx] = results
+
+    metrics.incr(
+        "metrics_layer.query",
+        tags={**logging_tags, "status": "success"},
+    )
+    return snuba_results
+
+
 def run_query(request: Request) -> Mapping[str, Any]:
     """
     Entrypoint for executing a metrics query in Snuba.
-
-    First iteration:
-    The purpose of this function is to eventually replace datasource.py::get_series().
-    As a first iteration, this function will only support single timeseries metric queries.
-    This means that for now, other queries such as total, formula, or meta queries
-    will not be supported. Additionally, the first iteration will only support
-    querying raw metrics (no derived). This means that each call to this function will only
-    resolve into a single request (and single entity) to the Snuba API.
     """
+    request, start, end = _setup_metrics_query(request)
```

Comment:
Maybe a dumb question:

Why is the implementation of `run_query` not just

```python3
def run_query(request: Request) -> Mapping[str, Any]:
    return bulk_run_query{[request])[0]
```

Reply:
Not a dumb question, something that didn't occur to me.